### ISSUE TITLE
IPC Channel Closed errors when running in CircleCI

### ISSUE THREAD

**Issue by Osmose** [NeuraLegion/cypress-har-generator#23]
We're running our integration tests in CircleCI using the `circleci/node:10-browsers` image. We don't have any other Cypress plugins configured; this was the first. We're running Cypress 3.8.0.

After adding the plugin to our build/test process, we noticed most, but not all, of our test runs were failing with a message like this:

```
The following error was thrown by a plugin. We've stopped running your tests because a plugin crashed.

Error [ERR_IPC_CHANNEL_CLOSED] [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at ChildProcess.target.send (internal/child_process.js:678:16)
    at Object.send (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/util.js:34:27)
    at /home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/index.js:82:28
    at /home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/util.js:77:16
    at Promise.cancellationExecute [as _execute] (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/debuggability.js:449:9)
    at Promise._resolveFromExecutor (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:518:18)
    at new Promise (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:103:10)
    at Object.wrapParentPromise (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/util.js:61:14)
    at Object._get:task:body (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/index.js:75:27)
    at Object.execute (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/index.js:128:38)
    at /home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/task.js:56:24
    at tryCatcher (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
    at /home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/catch_filter.js:17:41
    at tryCatcher (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromise0 (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:649:10)
    at Promise._settlePromises (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:725:18)
    at _drainQueueStep (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:93:12)
    at _drainQueue (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:86:9)
    at Async._drainQueues (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (/home/circleci/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:15:14)
    at processImmediate (internal/timers.js:439:21)
```

It's technically intermittent because we'd get occasional passes, and it's random when it happens in a run.

A smaller amount of runs are showing this message, although I don't think this is causing them to fail, just causing them to not have a saved HAR file:

```
🍑 Failed to save HAR: WebSocket is not open: readyState 3 (CLOSED)
```

I found a few issues on the main Cypress repo with similar messages but none seem related:
- https://github.com/cypress-io/cypress/issues/1968 (I checked with `npx native-modules` and didn't find any compiled modules being used during the integration tests)
- https://github.com/cypress-io/cypress/issues/1953 (our arguments are properly quoted)

Any help diagnosing the issue here would be appreciated. Thank you!

(Also I'm on vacation starting tomorrow so apologies in advance if I respond slowly over the next week or so.)

**Comment by ArtemDerevnjuk** [comment]
@Osmose thanks for the report)

I tried to reproduce that by running about simple 100 tests.

```javascript
describe('IPC Channel Closed errors when running in CircleCI #23', () => {
  it('should launch google.com', () => {
    cy.visit('https://google.com', {
      headers: {'Accept-Language': 'en-US'}
    });
  });

  it('should type a search request', () => {
    cy.get('input[name="q"]').type('Google');
    cy.get('form').submit();
  });
});
```

That's what got. 

```
<--- Last few GCs --->

[22784:0x1fb9fc531000]   714949 ms: Mark-sweep 4088.7 (4101.3) -> 4088.7 (4101.3) MB, 71.0 / 0.0 ms  (average mu = 0.120, current mu = 0.000) allocation failure scavenge might not succeed
[22784:0x1fb9fc531000]   715012 ms: Mark-sweep 4088.7 (4101.3) -> 4088.7 (4101.3) MB, 62.8 / 0.0 ms  (average mu = 0.068, current mu = 0.001) last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x0dfe158e0dd1 <JSObject>
    0: builtin exit frame: parse(this=0x0dfe158c9a09 <Object map = 0xbd8d4501569>,0x25b194280119 <Very long string[1890765]>,0x0dfe158c9a09 <Object map = 0xbd8d4501569>)

    1: /* anonymous */ [0x338a5aeec681] [/home/derevnjuk/Projects/CypressWebSocket/node_modules/chrome-remote-interface/lib/chrome.js:233] [bytecode=0x69715baee31 offset=12](this=0x338a5aed2f61 <EventEmitter map = 0x2e844f9c46c...


The following error was thrown by a plugin. We've stopped running your tests because a plugin crashed.

Error [ERR_IPC_CHANNEL_CLOSED] [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at ChildProcess.target.send (internal/child_process.js:678:16)
    at Object.send (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/util.js:34:27)
    at /home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/index.js:82:28
    at /home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/util.js:77:16
    at Promise.cancellationExecute [as _execute] (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/debuggability.js:449:9)
    at Promise._resolveFromExecutor (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:518:18)
    at new Promise (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:103:10)
    at Object.wrapParentPromise (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/util.js:61:14)
    at Object._get:task:body (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/index.js:75:27)
    at Object.execute (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/plugins/index.js:128:38)
    at /home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/lib/task.js:56:24
    at tryCatcher (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
    at /home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/catch_filter.js:17:41
    at tryCatcher (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromise0 (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:649:10)
    at Promise._settlePromises (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:725:18)
    at _drainQueueStep (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:93:12)
    at _drainQueue (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:86:9)
    at Async._drainQueues (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (/home/derevnjuk/.cache/Cypress/3.8.0/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:15:14)
    at processImmediate (internal/timers.js:439:21)
```

This error occurs when the memory allocated is less than the required memory to serialize HAR. Seems the plugin uses all available heap to keep in memory every request until the end of testing. By default the heap's limit in Node.js is 512MB, you can solve this issue by increasing `max-old-space-size` via `NODE_OPTIONS`.

I guess it'll be fixed by #22

**Comment by ArtemDerevnjuk** [comment]
Regarding `WebSocket is not open: readyState 3 (CLOSED)`. I assume it has to do with a reference to old Network instance in request objects. I tried to fix that, but unfortunately, I was faced with `No resource with given id was found` error. Seems Chrome discards resources after page navigation.
